sympy/core/function.py

```diff
@@ -646,14 +646,25 @@ def fdiff(self, argindex=1):
         """
         if not (1 <= argindex <= len(self.args)):
             raise ArgumentIndexError(self, argindex)
-        if not self.args[argindex - 1].is_Symbol:
-            # See issue 4624 and issue 4719 and issue 5600
-            arg_dummy = Dummy('xi_%i' % argindex)
-            arg_dummy.dummy_index = hash(self.args[argindex - 1])
-            return Subs(Derivative(
-                self.subs(self.args[argindex - 1], arg_dummy),
-                arg_dummy), arg_dummy, self.args[argindex - 1])
-        return Derivative(self, self.args[argindex - 1], evaluate=False)
+
+        if self.args[argindex - 1].is_Symbol:
+            use_dummy = False
+            for i in range(len(self.args)):
+                if i == argindex:
+                    continue
+                # See issue 8510
+                if self.args[argindex - 1] in self.args[i].free_symbols:
+                    use_dummy=True
+                    break
+            if not use_dummy:
+                return Derivative(self, self.args[argindex - 1], evaluate=False)
```

Comment:
You can use ``for else`` construct here. That can save you from declaring ``use_dummy``.